libr/cons/dietline.c

```diff
@@ -1040,19 +1056,25 @@ R_API const char *r_line_readline_cb_win(RLineReadCallback cb, void *user) {
 				gcomp++;
 			}
 			if (I.buffer.index < I.buffer.length) {
-				for (i = ++I.buffer.length; i > I.buffer.index; i--) {
-					I.buffer.data[i] = I.buffer.data[i - 1];
+				int size = r_str_utf8_charsize (buf);
+				I.buffer.length += size;
+				for (i = I.buffer.length; i - size + 1 > I.buffer.index; i--) {
+					I.buffer.data[i] = I.buffer.data[i - size];
+				}
+				while (size) {
+					I.buffer.data[i--] = buf[--size];
 				}
-				I.buffer.data[I.buffer.index] = buf[0];
 			} else {
-				I.buffer.data[I.buffer.length] = buf[0];
-				I.buffer.length++;
+				for (i = 0; i < R_LINE_BUFSIZE - 1 && buf[i]; i++) {
+					I.buffer.data[I.buffer.length + i] = buf[i];
```

Comment:
I think there will be an overflow of `I.buffer.data` in case when it already holds some data.